tools/pkg/engine_build_configs/lib/src/build_config_runner.dart

```diff
@@ -0,0 +1,553 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:async';
+import 'dart:convert';
+import 'dart:io' as io show Directory, Process;
+
+import 'package:path/path.dart' as p;
+import 'package:platform/platform.dart';
+import 'package:process_runner/process_runner.dart';
+
+import 'build_config.dart';
+
+/// The base clase for events generated by a command.
+sealed class RunnerEvent {
+  RunnerEvent(this.name, this.command, this.timestamp);
+
+  /// The name of the task or command.
+  final String name;
+
+  /// The command and its arguments.
+  final List<String> command;
+
+  /// When the event happened.
+  final DateTime timestamp;
+}
+
+/// A [RunnerEvent] representing the start of a command.
+final class RunnerStart extends RunnerEvent {
+  RunnerStart(super.name, super.command, super.timestamp);
+
+  @override
+  String toString() {
+    return '[${_timestamp(timestamp)}][$name]: STARTING';
+  }
+}
+
+/// A [RunnerEvent] representing the progress of a started command.
+final class RunnerProgress extends RunnerEvent {
+  RunnerProgress(
+    super.name, super.command, super.timestamp,
+    this.what, this.percent, this.done,
+  );
+
+  /// What a command is currently working on, for example a build target or
+  /// the name of a test.
+  final String what;
+
+  /// How close is the task to being completed, for example the proportion of
+  /// build targets that have finished building.
+  final double percent;
+
+  /// Whether the command is finished and this is the final progress event.
+  final bool done;
+
+  @override
+  String toString() {
+    final String ts = '[${_timestamp(timestamp)}]';
+    final String pct = '${percent.toStringAsFixed(1)}%';
+    return '$ts[$name]: $pct $what';
```

Comment:
The percentages feel slower than ninja's `<completed>/<total>` output. My computer might display `68.9%` for a while even though it's actually completed several steps.

Could you expand on why you chose percentages instead of raw numbers?

Reply:
Didn't think about this too deeply, but it does make sense for the progress object to include the number of completed steps and the total number of steps so that the client of the API can do whatever it likes with them.